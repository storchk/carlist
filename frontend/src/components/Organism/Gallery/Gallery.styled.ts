import styled, { css } from 'styled-components'

export const StyledGallery = styled.div`
  display: flex;
  flex-direction: column;
`
export const StyledHeroImage = styled.img(
  ({ theme }) => css`
    width: 100%;
    margin-bottom: ${theme.spacing.md};
    border-radius: ${theme.border.radius};
  `
)

export const StyledPreviewImageList = styled.ul(
  ({ theme }) => css`
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    gap: ${theme.spacing.sm} ${theme.spacing.xs};

    &::after {
      content: '';
      flex: auto;
    }
  `
)

export const StyledPreviewImageListItem = styled.li<{ $isActive: boolean }>(
  ({ theme, $isActive }) => css`
    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden;
    border-radius: ${theme.border.radius};
    transition: all 0.5s cubic-bezier(0.2, 0, 0.05, 1);

    width: calc(20% - ${theme.spacing.xs});
    height: calc(20% - ${theme.spacing.xs});

    ${$isActive
      ? css`
          border: 4px solid ${theme.colors.primary};
        `
      : css`
          border: 4px solid transparent;
        `}

    &:hover {
      cursor: pointer;
      transform: scale(1.02);
    }
  `
)

export const StyledPreviewImage = styled.img`
  width: 100%;
  height: -webkit-fill-available;
`
