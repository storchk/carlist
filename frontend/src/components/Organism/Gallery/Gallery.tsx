import { useState } from 'react'

import {
  StyledGallery,
  StyledHeroImage,
  StyledPreviewImage,
  StyledPreviewImageList,
  StyledPreviewImageListItem,
} from './Gallery.styled'
import type { GalleryProps } from './Gallery.types'

export const Gallery = ({ images }: GalleryProps): JSX.Element => {
  const [activeImage, setActiveImage] = useState(images[0])
  return (
    <StyledGallery>
      <StyledHeroImage src={activeImage} alt="" data-test="gallery-active-image" />
      <StyledPreviewImageList>
        {images.map((image, index) => {
          return (
            <StyledPreviewImageListItem
              key={`gallery-image-${index}`}
              $isActive={image === activeImage}
              onClick={() => setActiveImage(image)}
            >
              <StyledPreviewImage src={image} alt="" loading="lazy" />
            </StyledPreviewImageListItem>
          )
        })}
      </StyledPreviewImageList>
    </StyledGallery>
  )
}
