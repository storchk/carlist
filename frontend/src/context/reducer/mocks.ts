import type { Car } from '@/graphql'
export const mockedCar1: Car = {
  brand: 'Volkswagen',
  category: 'LIMOUSINE',
  color: 'WHITE',
  drivetrain: {
    cc: '999 ccm',
    consumption: '5.2 LITER',
    fuel: 'Benzin',
    transmissionType: 'Schaltgetriebe',
  },
  id: '32725100-3151-5348-871a-faee7be1f699',
  media: [
    {
      url: 'https://dqg4c1i67ln7w.cloudfront.net/photos/edited/0dc7206c81b56879b7c42a3ecca3ad4e/image_0PU067451_20.jpg',
    },
  ],
  model: 'Polo',
  performance: '70 PS',
  price: {
    gross: 22395,
  },
  technicalFeatures: {
    drive: 'NO INFORMATION',
  },
  vehicleHistory: {
    registrationDate: '06.2023',
  },
  vehicleType: {
    condition: 'USED',
  },
}

export const mockedCar2: Car = {
  brand: 'Opel',
  category: 'OFFROAD',
  color: 'RED',
  drivetrain: {
    cc: '1199 ccm',
    consumption: '4.6 LITER',
    fuel: 'Benzin',
    transmissionType: 'Schaltgetriebe',
  },
  id: 'd5ae265a-5a44-5e4b-b25b-1b8825072e65',
  media: [
    {
      url: 'https://dqg4c1i67ln7w.cloudfront.net/photos/edited/91597c07abbab7d535ac280cf0ef6f39/image_04.jpg',
    },
    {
      url: 'https://dqg4c1i67ln7w.cloudfront.net/photos/edited/91597c07abbab7d535ac280cf0ef6f39/image_01.jpg',
    },
  ],
  model: 'Crossland',
  performance: '60 PS',
  price: {
    gross: 23980,
  },
  technicalFeatures: {
    drive: null,
  },
  vehicleHistory: {
    registrationDate: '09.2022',
  },
  vehicleType: {
    condition: 'USED',
  },
}
