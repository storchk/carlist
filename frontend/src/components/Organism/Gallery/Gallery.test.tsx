import { axe } from 'jest-axe'
import { testing } from '../../../testing'
import { Gallery } from './Gallery'
import { mockedCar1 } from '@/mocks'
const { render, screen, renderer } = testing
import userEvent from '@testing-library/user-event'

const mockedImages = mockedCar1.media.map(image => image.url)

describe('Organism:Gallery', () => {
  it('renders the correct number of images', () => {
    render(<Gallery images={mockedImages} />)
    const imageElements = screen.getAllByRole('img')
    expect(imageElements).toHaveLength(mockedImages.length + 1)
  })

  it('should change hero image', () => {
    render(<Gallery images={mockedImages} />)

    const imageElements = screen.getAllByRole('img')
    expect(screen.getByTestId('gallery-active-image')).toHaveAttribute('src', mockedImages[0])
    userEvent.click(imageElements[1])
    expect(screen.getByTestId('gallery-active-image')).toHaveAttribute('src', mockedImages[1])
  })

  it('snapshot', () => {
    const view = renderer(<Gallery images={mockedImages} />)
    expect(view).toMatchSnapshot()
  })

  it('a11y - No violations detected', async () => {
    const { container } = render(<Gallery images={mockedImages} />)

    await expect(axe(container)).resolves.toHaveNoViolations()
  })
})
