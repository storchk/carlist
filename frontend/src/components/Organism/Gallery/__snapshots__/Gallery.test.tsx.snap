// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organism:Gallery snapshot 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
}

.c1 {
  width: 100%;
  margin-bottom: 24px;
  border-radius: 8px;
}

.c2 {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  gap: 16px 8px;
}

.c2::after {
  content: '';
  flex: auto;
}

.c3 {
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  border-radius: 8px;
  transition: all 0.5s cubic-bezier(0.2, 0, 0.05, 1);
  width: calc(20% - 8px);
  height: calc(20% - 8px);
  border: 4px solid #f7b628;
}

.c3:hover {
  cursor: pointer;
  transform: scale(1.02);
}

.c4 {
  width: 100%;
  height: -webkit-fill-available;
}

<div
  className="c0"
>
  <img
    alt=""
    className="c1"
    data-test="gallery-active-image"
    src="https://dqg4c1i67ln7w.cloudfront.net/photos/edited/0dc7206c81b56879b7c42a3ecca3ad4e/image_0PU067451_20.jpg"
  />
  <ul
    className="c2"
  >
    <li
      className="c3"
      onClick={[Function]}
    >
      <img
        alt=""
        className="c4"
        loading="lazy"
        src="https://dqg4c1i67ln7w.cloudfront.net/photos/edited/0dc7206c81b56879b7c42a3ecca3ad4e/image_0PU067451_20.jpg"
      />
    </li>
  </ul>
</div>
`;
